core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -333,6 +347,122 @@ public long nextFetchOffset() {
         }
     }
 
+    void updateOffsetsOnLsoMovement() {
+        long logStartOffset = offsetForEarliestTimestamp();
+        lock.writeLock().lock();
+        try {
+            if (logStartOffset > startOffset) {
+                log.debug("Updating start offset for share partition: {}-{} from: {} to: {} since LSO has moved to: {}",
+                        groupId, topicIdPartition, startOffset, logStartOffset, logStartOffset);
+                if (cachedState.isEmpty()) {
+                    startOffset = logStartOffset;
+                    endOffset = logStartOffset;
+                } else {
```

Comment:
nit: just personal preference: I don't find the code is very readable with a lot of if/else rather it's much better with `if` followed by a `return` is possible. Please feel free to ignore the comment if you don't prefer it this way.

```
if (cachedState.isEmpty()) {
           startOffset = logStartOffset;
           endOffset = logStartOffset;
           return;
}
List<PersisterStateBatch> stateBatches = new ArrayList<>();
...